src/main/scala/rise/core/DSL/infer.scala

```diff
@@ -123,31 +78,49 @@ object infer {
 
   private val genType : Expr => Type = e => if (e.t == TypePlaceholder) freshTypeIdentifier else e.t
 
-  private val constrainTypes : Map[String, Type] => Expr => (Expr, Seq[Constraint], Solution) = env => {
+  private val collectPreserve = new PureAccumulatorTraversal[Seq[Kind.Identifier]] {
+    override val accumulator = SeqMonoid
```

Comment:
I think that this should really be a `HashMap` for performance reasons? Ideally indentifiers should also not be string-based in the future (https://github.com/rise-lang/shine/issues/137).

Reply:
I put them in a `HashSet` -- and yes, when identifiers become integer based so does this.